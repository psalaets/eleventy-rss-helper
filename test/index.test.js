const MockDate = require('mockdate');
const createRssFeed = require('..');

describe('xml', () => {
  beforeEach(() => MockDate.set('August 17, 2019 11:43:02'));
  afterEach(() => MockDate.reset());

  test('minimal', () => {
    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        return {
          title: 'My Blog',
          feed_url: 'https://myblog.com/feed.xml',
          site_url: 'https://myblog.com'
        };
      },
      items() {
        return [1, 2];
      },
      itemOptions(number) {
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com/${number}`,
          date: `2019-0${number}-01`
        };
      }
    });

    const xml = renderXml(cls);

    expect(xml).toMatchSnapshot();
  });

  test('everything on', () => {
    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        return {
          title: 'My Blog',
          description: 'Posts from my blog',
          generator: 'my rss generator',
          feed_url: 'https://myblog.com/feed.xml',
          site_url: 'https://myblog.com',
          image_url: 'https://myblog.com/cover-image.png',
          docs: 'https://myblog.com/feed-docs',
          managingEditor: 'First McName',
          webMaster: 'Name Last',
          copyright: '2019',
          language: 'en-US',
          categories: ['sci-fi', 'technology'],
          pubDate: '2019-08-18',
          ttl: 24 * 60,
          hub: 'https://myblog.com/hub',
          custom_namespaces: {},
          custom_elements: []
        };
      },
      items() {
        return [1, 2];
      },
      itemOptions(number) {
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com/${number}`,
          guid: `https://myblog.com/guid-${number}`,
          categories: ['poetry', 'music'],
          author: 'Jack Nimble',
          date: `2019-0${number}-01`,
          lat: 30,
          long: 25,
          custom_elements: [],
          enclosure: {
            url: `https://myblog.com/episode-${number}`,
            size: 1024 * number,
            type: 'image/png'
          }
        };
      }
    });

    const xml = renderXml(cls);

    expect(xml).toMatchSnapshot();
  });

  test('normalizes urls in feedOptions', () => {
    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        // double slash after hostname
        return {
          title: 'My Blog',
          feed_url: 'https://myblog.com//feed.xml',
          site_url: 'https://myblog.com//',
          image_url: 'https://myblog.com//cover-image.png',
          docs: 'https://myblog.com//feed-docs',
          hub: 'https://myblog.com//hub'
        };
      },
      items() {
        return [1, 2];
      },
      itemOptions(number) {
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com/${number}`,
          date: `2019-0${number}-01`
        };
      }
    });

    const xml = renderXml(cls);

    expect(xml).toMatchSnapshot();
  });

  test('normalizes urls in itemOptions', () => {
    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        return {
          title: 'My Blog',
          feed_url: 'https://myblog.com/feed.xml',
          site_url: 'https://myblog.com'
        };
      },
      items() {
        return [1, 2];
      },
      itemOptions(number) {
        // double slash after hostname
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com//${number}`,
          date: `2019-0${number}-01`,
          enclosure: {
            url: `https://myblog.com//episode-${number}`,
            size: 1024 * number,
            type: 'image/png'
          }
        };
      }
    });

    const xml = renderXml(cls);

    expect(xml).toMatchSnapshot();
  });
});

function renderXml(cls) {
  const instance = new cls();
  const data = instance.data();
  return instance.render({
    ...data,
    collections: {
      post: []
    }
  });
}

describe('options error checking', () => {
  test('throws Error when permalink not included', () => {
    expect(() => {
      createRssFeed({
        //permalink: '/feed.xml',
        feedOptions() {
          return {
            title: 'My Blog',
            feed_url: 'https://myblog.com/feed.xml',
            site_url: 'https://myblog.com'
          };
        },
        items() {
          return [1, 2];
        },
        itemOptions(number) {
          return {
            title: `Item ${number}`,
            description: `This is item ${number}`,
            url: `https://myblog.com/${number}`,
            date: `2019-0${number}-01`
          };
        }
      });
    }).toThrow(Error);
  });

  test('throws Error when feedOptions() not included', () => {
    expect(() => {
      createRssFeed({
        permalink: '/feed.xml',
        // feedOptions() {
        //   return {
        //     title: 'My Blog',
        //     feed_url: 'https://myblog.com/feed.xml',
        //     site_url: 'https://myblog.com'
        //   };
        // },
        items() {
          return [1, 2];
        },
        itemOptions(number) {
          return {
            title: `Item ${number}`,
            description: `This is item ${number}`,
            url: `https://myblog.com/${number}`,
            date: `2019-0${number}-01`
          };
        }
      });
    }).toThrow(Error);
  });

  test('throws Error when items() not included', () => {
    expect(() => {
      createRssFeed({
        permalink: '/feed.xml',
        feedOptions() {
          return {
            title: 'My Blog',
            feed_url: 'https://myblog.com/feed.xml',
            site_url: 'https://myblog.com'
          };
        },
        // items() {
        //   return [1, 2];
        // },
        itemOptions(number) {
          return {
            title: `Item ${number}`,
            description: `This is item ${number}`,
            url: `https://myblog.com/${number}`,
            date: `2019-0${number}-01`
          };
        }
      });
    }).toThrow(Error);
  });

  test('throws Error when itemOptions() not included', () => {
    expect(() => {
      createRssFeed({
        permalink: '/feed.xml',
        feedOptions() {
          return {
            title: 'My Blog',
            feed_url: 'https://myblog.com/feed.xml',
            site_url: 'https://myblog.com'
          };
        },
        items() {
          return [1, 2];
        },
        // itemOptions(number) {
        //   return {
        //     title: `Item ${number}`,
        //     description: `This is item ${number}`,
        //     url: `https://myblog.com/${number}`,
        //     date: `2019-0${number}-01`
        //   };
        // }
      });
    }).toThrow(Error);
  });
});

describe('calling user-supplied functions during render()', () => {
  test('passes data object to feedOptions()', () => {
    const mockFeedOptions = jest.fn();
    mockFeedOptions.mockReturnValue({});

    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions: mockFeedOptions,
      items() {
        return [1, 2];
      },
      itemOptions(number) {
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com/${number}`,
          date: `2019-0${number}-01`
        };
      }
    });

    const instance = new cls();
    const eleventyDataObject = {
      collections: {}
    };

    instance.data();
    instance.render(eleventyDataObject);

    expect(mockFeedOptions).toHaveBeenCalledWith(eleventyDataObject);
  });

  test('passes collections and data object to items()', () => {
    const mockItems = jest.fn();
    mockItems.mockReturnValue([]);

    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        return {
          title: 'My Blog',
          feed_url: 'https://myblog.com/feed.xml',
          site_url: 'https://myblog.com'
        };
      },
      items: mockItems,
      itemOptions(number) {
        return {
          title: `Item ${number}`,
          description: `This is item ${number}`,
          url: `https://myblog.com/${number}`,
          date: `2019-0${number}-01`
        };
      }
    });

    const instance = new cls();
    const eleventyCollectionsObject = {};
    const eleventyDataObject = {
      collections: eleventyCollectionsObject
    };

    instance.data();
    instance.render(eleventyDataObject);

    expect(mockItems).toHaveBeenCalledWith(eleventyCollectionsObject, eleventyDataObject);
  });

  test('passes items and data object to itemOptions', () => {
    const mockItemOptions = jest.fn();
    mockItemOptions.mockReturnValue({});

    const cls = createRssFeed({
      permalink: '/feed.xml',
      feedOptions() {
        return {
          title: 'My Blog',
          feed_url: 'https://myblog.com/feed.xml',
          site_url: 'https://myblog.com'
        };
      },
      items() {
        return [1, 2];
      },
      itemOptions: mockItemOptions
    });

    const instance = new cls();
    const eleventyDataObject = {
      collections: {}
    };

    instance.data();
    instance.render(eleventyDataObject);

    expect(mockItemOptions).toHaveBeenNthCalledWith(1, 1, eleventyDataObject);
    expect(mockItemOptions).toHaveBeenNthCalledWith(2, 2, eleventyDataObject);
  });
});
