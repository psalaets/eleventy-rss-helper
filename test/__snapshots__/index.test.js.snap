// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xml everything on 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><rss xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\" xmlns:content=\\"http://purl.org/rss/1.0/modules/content/\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\" version=\\"2.0\\" xmlns:geo=\\"http://www.w3.org/2003/01/geo/wgs84_pos#\\"><channel><title><![CDATA[My Blog]]></title><description><![CDATA[Posts from my blog]]></description><link>https://myblog.com</link><image><url>https://myblog.com/cover-image.png</url><title>My Blog</title><link>https://myblog.com</link></image><generator>my rss generator</generator><lastBuildDate>Sat, 17 Aug 2019 15:43:02 GMT</lastBuildDate><atom:link href=\\"https://myblog.com/feed.xml\\" rel=\\"self\\" type=\\"application/rss+xml\\"/><pubDate>Sun, 18 Aug 2019 00:00:00 GMT</pubDate><copyright><![CDATA[2019]]></copyright><language><![CDATA[en-US]]></language><managingEditor><![CDATA[First McName]]></managingEditor><webMaster><![CDATA[Name Last]]></webMaster><docs>https://myblog.com/feed-docs</docs><ttl>1440</ttl><atom:link href=\\"https://myblog.com/hub\\" rel=\\"hub\\"/><category><![CDATA[sci-fi]]></category><category><![CDATA[technology]]></category><item><title><![CDATA[Item 1]]></title><description><![CDATA[This is item 1]]></description><link>https://myblog.com/1</link><guid isPermaLink=\\"false\\">https://myblog.com/guid-1</guid><category><![CDATA[poetry]]></category><category><![CDATA[music]]></category><dc:creator><![CDATA[Jack Nimble]]></dc:creator><pubDate>Tue, 01 Jan 2019 00:00:00 GMT</pubDate><geo:lat>30</geo:lat><geo:long>25</geo:long><enclosure url=\\"https://myblog.com/episode-1\\" length=\\"1024\\" type=\\"image/png\\"/></item><item><title><![CDATA[Item 2]]></title><description><![CDATA[This is item 2]]></description><link>https://myblog.com/2</link><guid isPermaLink=\\"false\\">https://myblog.com/guid-2</guid><category><![CDATA[poetry]]></category><category><![CDATA[music]]></category><dc:creator><![CDATA[Jack Nimble]]></dc:creator><pubDate>Fri, 01 Feb 2019 00:00:00 GMT</pubDate><geo:lat>30</geo:lat><geo:long>25</geo:long><enclosure url=\\"https://myblog.com/episode-2\\" length=\\"2048\\" type=\\"image/png\\"/></item></channel></rss>"`;

exports[`xml minimal 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><rss xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\" xmlns:content=\\"http://purl.org/rss/1.0/modules/content/\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\" version=\\"2.0\\"><channel><title><![CDATA[My Blog]]></title><description><![CDATA[My Blog]]></description><link>https://myblog.com</link><generator>RSS for Node</generator><lastBuildDate>Sat, 17 Aug 2019 15:43:02 GMT</lastBuildDate><atom:link href=\\"https://myblog.com/feed.xml\\" rel=\\"self\\" type=\\"application/rss+xml\\"/><item><title><![CDATA[Item 1]]></title><description><![CDATA[This is item 1]]></description><link>https://myblog.com/1</link><guid isPermaLink=\\"true\\">https://myblog.com/1</guid><pubDate>Tue, 01 Jan 2019 00:00:00 GMT</pubDate></item><item><title><![CDATA[Item 2]]></title><description><![CDATA[This is item 2]]></description><link>https://myblog.com/2</link><guid isPermaLink=\\"true\\">https://myblog.com/2</guid><pubDate>Fri, 01 Feb 2019 00:00:00 GMT</pubDate></item></channel></rss>"`;

exports[`xml normalizes urls in feedOptions 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><rss xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\" xmlns:content=\\"http://purl.org/rss/1.0/modules/content/\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\" version=\\"2.0\\"><channel><title><![CDATA[My Blog]]></title><description><![CDATA[My Blog]]></description><link>https://myblog.com</link><image><url>https://myblog.com/cover-image.png</url><title>My Blog</title><link>https://myblog.com</link></image><generator>RSS for Node</generator><lastBuildDate>Sat, 17 Aug 2019 15:43:02 GMT</lastBuildDate><atom:link href=\\"https://myblog.com/feed.xml\\" rel=\\"self\\" type=\\"application/rss+xml\\"/><docs>https://myblog.com/feed-docs</docs><atom:link href=\\"https://myblog.com/hub\\" rel=\\"hub\\"/><item><title><![CDATA[Item 1]]></title><description><![CDATA[This is item 1]]></description><link>https://myblog.com/1</link><guid isPermaLink=\\"true\\">https://myblog.com/1</guid><pubDate>Tue, 01 Jan 2019 00:00:00 GMT</pubDate></item><item><title><![CDATA[Item 2]]></title><description><![CDATA[This is item 2]]></description><link>https://myblog.com/2</link><guid isPermaLink=\\"true\\">https://myblog.com/2</guid><pubDate>Fri, 01 Feb 2019 00:00:00 GMT</pubDate></item></channel></rss>"`;

exports[`xml normalizes urls in itemOptions 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><rss xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\" xmlns:content=\\"http://purl.org/rss/1.0/modules/content/\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\" version=\\"2.0\\"><channel><title><![CDATA[My Blog]]></title><description><![CDATA[My Blog]]></description><link>https://myblog.com</link><generator>RSS for Node</generator><lastBuildDate>Sat, 17 Aug 2019 15:43:02 GMT</lastBuildDate><atom:link href=\\"https://myblog.com/feed.xml\\" rel=\\"self\\" type=\\"application/rss+xml\\"/><item><title><![CDATA[Item 1]]></title><description><![CDATA[This is item 1]]></description><link>https://myblog.com/1</link><guid isPermaLink=\\"true\\">https://myblog.com/1</guid><pubDate>Tue, 01 Jan 2019 00:00:00 GMT</pubDate><enclosure url=\\"https://myblog.com//episode-1\\" length=\\"1024\\" type=\\"image/png\\"/></item><item><title><![CDATA[Item 2]]></title><description><![CDATA[This is item 2]]></description><link>https://myblog.com/2</link><guid isPermaLink=\\"true\\">https://myblog.com/2</guid><pubDate>Fri, 01 Feb 2019 00:00:00 GMT</pubDate><enclosure url=\\"https://myblog.com//episode-2\\" length=\\"2048\\" type=\\"image/png\\"/></item></channel></rss>"`;
